File: data_processor.cpp
#include <iostream>
#include <cstdlib>
#include <chrono>
#include <thread>
#include <unistd.h>
#include "json.hpp"
#include "mqtt/client.h"
#include <string>
#include <vector>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <deque>
#include <cmath>
#include <ctime>

#define QOS 1
#define BROKER_ADDRESS "tcp://localhost:1883"
#define GRAPHITE_HOST "127.0.0.1"
#define GRAPHITE_PORT 2003

// CÁLCULO E CONVERSÃO -------------------------------------------------------------------------------------------

std::time_t string_to_time_t(const std::string& time_string) {
    std::tm tm = {};
    std::istringstream ss(time_string);
    ss >> std::get_time(&tm, "%Y-%m-%dT%H:%M:%S");
    return std::mktime(&tm);
}

std::tm string_to_tm(const std::string& timestamp) {
    std::tm tm = {};
    std::istringstream ss(timestamp);
    ss >> std::get_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
    return tm;
}

// Defina o tamanho da janela para a média móvel
const size_t MOVING_AVERAGE_WINDOW = 5; // Por exemplo, uma janela de tamanho 5

// Função para calcular a média móvel de um conjunto de valores
double calculateMovingAverage(const std::deque<double>& values) {
    if (values.empty()) {
        return 0.0; // Se a fila estiver vazia, retorna 0 como média móvel
    }

    double sum = 0.0;
    for (const auto& value : values) {
        sum += value;
    }

    return sum / static_cast<double>(values.size());
}


// Função para calcular o Z-score de um valor em relação a um conjunto de dados
double calculateZScore(double value, const std::deque<double>& values) {
    if (values.empty()) {
        return 0.0; // Se a fila estiver vazia, retorna 0 como Z-score
    }

    double mean = calculateMovingAverage(values);
    double variance = 0.0;

    for (const auto& val : values) {
        variance += std::pow(val - mean, 2);
    }

    variance /= static_cast<double>(values.size());
    double stdDeviation = std::sqrt(variance);

    if (stdDeviation == 0.0) {
        return 0.0; // Retorna 0 se o desvio padrão for zero para evitar divisão por zero
    }

    return (value - mean) / stdDeviation;
}

// Função para calcular a tendência usando regressão linear simples
double calculateTrend(const std::deque<double>& values) {
    size_t n = values.size();
    double sumX = 0.0, sumY = 0.0, sumXY = 0.0, sumXX = 0.0;

    if (n < 2) {
        return 0.0;
    }

    for (size_t i = 0; i < n; ++i) {
        sumX += i + 1;
        sumY += values[i];
        sumXY += (i + 1) * values[i];
        sumXX += (i + 1) * (i + 1);
    }

    double slope = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);

    return slope;
}

// POSTAR MÉTRICA ------------------------------------------------------------------------------------------

int post_metric(const std::string& machine_id, const std::string& sensor_id, const std::string& timestamp_str, const double value) {
    int graphite_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (graphite_socket == -1) {
        std::cerr << "Error: Failed to create socket\n";
        return -1; // Retorna um código de erro
    }

    struct sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(GRAPHITE_PORT);
    if (inet_pton(AF_INET, GRAPHITE_HOST, &server_addr.sin_addr) <= 0) {
        std::cerr << "Error: Invalid address or address not supported\n";
        close(graphite_socket);
        return -1; // Retorna um código de erro
    }

    if (connect(graphite_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) == -1) {
        std::cerr << "Error: Failed to connect to Graphite\n";
        close(graphite_socket);
        return -1; // Retorna um código de erro
    }

    std::string graphite_topic = "machines." + machine_id + "." + sensor_id;
    std::stringstream metric_stream;
    metric_stream << graphite_topic << " " << value << " " << string_to_time_t(timestamp_str) << "\n";

    std::string metric_data = metric_stream.str();
    ssize_t sent_bytes = send(graphite_socket, metric_data.c_str(), metric_data.size(), 0);
    if (sent_bytes == -1) {
        std::cerr << "Error: Failed to send metric data for topic: " << graphite_topic << " with value: " << value << "\n";
        close(graphite_socket);
        return -1; // Retorna um código de erro
    } else {
        //std::cout << "Metric published to Graphite successfully for topic: " << graphite_topic << " with value: " << value << "\n";
    }

    close(graphite_socket);
    return 0; // Retorna sucesso
}

// PROCESSAMENTO DE DADOS ---------------------------------------------------------------------------------------

void process_sensor_data(const std::string& machine_id, const std::string& sensor_id, const std::string& timestamp, 
const double value, std::deque<double>& sensorData, std::deque<std::string>& sensorTimestamp) {

    std::cout << "\n\n" << "--------------------->    Análise de dados para o sensor " << sensor_id << "   <---------------------\n";
    std::tm time = string_to_tm(timestamp);
    std::cout << "Data: " << std::put_time(&time, "%d/%m/%Y, Hora: %H:%M:%S") << std::endl;
     std::cout << "ID da máquina: " << machine_id << std::endl;

    // Coleta de dados do sensor
    sensorData.push_back(value);
    sensorTimestamp.push_back(timestamp);

    // Mantenha o tamanho máximo da janela
    if (sensorData.size() > MOVING_AVERAGE_WINDOW) {
        sensorData.pop_front(); // Remova o valor mais antigo se exceder o tamanho da janela
        sensorTimestamp.pop_front();
    }

    // Calcular a média móvel do sensor
    if (sensorData.size() > 0) {
        double movingAverage = 0.0;
        for (const auto& data : sensorData) {
            movingAverage += data;
        }
        movingAverage /= sensorData.size();

        std::cout << "Média móvel do uso de " << sensor_id << ": " << movingAverage << std::endl;
        post_metric(machine_id, sensor_id + "." + sensor_id + "_moving_average", timestamp, movingAverage);

        // Detectar outliers usando Z-score
        double zScore = calculateZScore(value, sensorData);
        double zScoreThreshold = 1.0; // Defina o limite de Z-score para considerar um ponto como outlier

        if (std::abs(zScore) > zScoreThreshold) {
            // Se o valor atual for um outlier
            std::cout << "[ALARME] Outlier detectado: " << value << std::endl;
            post_metric(machine_id, "alarms." + sensor_id + "_outlier", timestamp, 1);
        } else {
            // Se não for um outlier
            std::cout << "Uso normal de " << sensor_id << ": " << value << std::endl;
        }
        // Calcular a tendência dos valores
        double trend = calculateTrend(sensorData);
        std::cout << "Tendência do uso de " << sensor_id << ": " << trend << std::endl;
        post_metric(machine_id, sensor_id + "." + sensor_id + "_trend", timestamp, trend);
        std::cout << "----------------------------------------------------------------------------------------------\n";
    }
}

void process_sensor_alarm(const std::string& machine_id, const std::string& sensor_id, const std::deque<std::string>* sensorTimestamps) {
    const int expected_interval_seconds = 30; // intervalo de chegada esperado
    const int max_expected_delay = expected_interval_seconds; // máximo de atraso esperado para gerar um alarme

    if (sensorTimestamps != nullptr && !sensorTimestamps->empty()) {
        std::string last_timestamp = sensorTimestamps->back();
        std::time_t last_time = string_to_time_t(last_timestamp);
        std::time_t current_time = std::time(nullptr);
        int seconds_since_last_timestamp = current_time - last_time;
        auto now = std::chrono::system_clock::now();
        std::time_t now_c = std::chrono::system_clock::to_time_t(now);
        std::tm* now_tm = std::gmtime(&now_c);
        std::stringstream ss;
        ss << std::put_time(now_tm, "%FT%TZ");
        std::string timestamp = ss.str();

        if (seconds_since_last_timestamp > max_expected_delay) {
            // Gerar alarme se o atraso for maior do que o esperado
            std::cout << "[ALARME] Dados do sensor " << sensor_id << " da máquina " << machine_id << " não foram recebidos por mais de 10 períodos de tempo previstos.\n";
            post_metric(machine_id, "alarms.inactive_" + sensor_id, timestamp, 1);
        }
    }
}

std::deque<std::string> cpuUsageTimestamps;
void process_sensor_data_cpu(const std::string& machine_id, const std::string& sensor_id, const std::string& timestamp, const double value) {
    static std::deque<double> cpuUsageData; // Use uma deque para armazenar valores anteriores
    process_sensor_data(machine_id, sensor_id, timestamp, value, cpuUsageData, cpuUsageTimestamps);
}

std::deque<std::string> memUsageTimestamps;
void process_sensor_data_mem(const std::string& machine_id, const std::string& sensor_id, const std::string& timestamp, const double value) {
    static std::deque<double> memUsageData; // Use uma deque para armazenar valores anteriores
    process_sensor_data(machine_id, sensor_id, timestamp, value, memUsageData, memUsageTimestamps);
}

std::vector<std::string> split(const std::string &str, char delim) {
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(str);
    while (std::getline(tokenStream, token, delim)) {
        tokens.push_back(token);
    }
    return tokens;
}


std::mutex mtx;
std::vector<std::string> monitored_sensors;

void check_sensor_inactivity(const std::string& machine_id, const std::string& sensor_id, const std::deque<std::string>* sensorTimestamps) {
    // Função para verificar a inatividade do sensor
    while (true) {
        {
            std::lock_guard<std::mutex> lock(mtx);
            process_sensor_alarm(machine_id, sensor_id, sensorTimestamps);
        }
        std::this_thread::sleep_for(std::chrono::seconds(20));
    }
}

void add_monitored_sensor(const std::string& sensor_id) {
    std::lock_guard<std::mutex> lock(mtx);
    monitored_sensors.push_back(sensor_id);
}

bool is_sensor_monitored(const std::string& sensor_id) {
    std::lock_guard<std::mutex> lock(mtx);
    return std::find(monitored_sensors.begin(), monitored_sensors.end(), sensor_id) != monitored_sensors.end();
}

void process_message(const std::string& machine_id, const std::string& sensor_id, const std::deque<std::string>* sensorTimestamps) {
    if (!is_sensor_monitored(sensor_id)) {
        std::thread sensor_thread(check_sensor_inactivity, machine_id, sensor_id, sensorTimestamps);
        sensor_thread.detach();
        add_monitored_sensor(sensor_id);
    }
}

// MAIN ---------------------------------------------------------------------------------------------------------

int main(int argc, char* argv[]) {
    std::string clientId = "clientId";
    mqtt::async_client client(BROKER_ADDRESS, clientId);

    // Create an MQTT callback.
    class callback : public virtual mqtt::callback {
    public:

        void message_arrived(mqtt::const_message_ptr msg) override {
            auto j = nlohmann::json::parse(msg->get_payload());

            std::string topic = msg->get_topic();
            auto topic_parts = split(topic, '/');
            std::string machine_id = topic_parts[2];
            std::string sensor_id = topic_parts[3];

            std::string timestamp = j["timestamp"];
            double value = j["value"];

            post_metric(machine_id, sensor_id + "." + sensor_id, timestamp, value);
            if (sensor_id == "cpu_usage") {
                process_sensor_data_cpu(machine_id, sensor_id, timestamp, value);
                process_message(machine_id, sensor_id, &cpuUsageTimestamps);
            } else {
                 process_sensor_data_mem(machine_id, sensor_id, timestamp, value);
                 process_message(machine_id, sensor_id, &memUsageTimestamps);
            }
        }
    };

    callback cb;
    client.set_callback(cb);
    // Connect to the MQTT broker.
    mqtt::connect_options connOpts;
    connOpts.set_keep_alive_interval(20);
    connOpts.set_clean_session(true);

    try {
        client.connect(connOpts)->wait();
        client.subscribe("/sensors/#", QOS);
        std::cout << "Subscrided\n";
    } catch (mqtt::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return EXIT_SUCCESS;
}

File: data_processor.py
import json
import time
import threading
from datetime import datetime, timezone
import numpy as np
import paho.mqtt.client as mqtt
from collections import deque
import psycopg2
from dateutil import parser
from typing import Deque

QOS = 1
BROKER_ADDRESS = "localhost"
MOVING_AVERAGE_WINDOW = 5

# Definição de cores ANSI para o console
RESET = "\033[0m"
RED = "\033[31m"

# Configuração do PostgreSQL
DB_HOST = "localhost"
DB_NAME = "postgres"
DB_USER = "postgres"
DB_PASSWORD = "1234"

def connect_to_db():
    return psycopg2.connect(
        host=DB_HOST,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD
    )

def create_table_if_not_exists():
    conn = connect_to_db()
    cursor = conn.cursor()
    create_table_query = '''
    CREATE TABLE IF NOT EXISTS metrics (
        id SERIAL PRIMARY KEY,
        machine_id VARCHAR(255) NOT NULL,
        sensor_id VARCHAR(255) NOT NULL,
        timestamp TIMESTAMPTZ NOT NULL,
        value DOUBLE PRECISION NOT NULL
    )
    '''
    cursor.execute(create_table_query)
    conn.commit()
    cursor.close()
    conn.close()

def post_metric(machine_id, sensor_id, timestamp_str, value):
    conn = connect_to_db()
    cursor = conn.cursor()
    
    insert_query = '''
    INSERT INTO metrics (machine_id, sensor_id, timestamp, value)
    VALUES (%s, %s, %s, %s)
    '''
    timestamp = parser.parse(timestamp_str).replace(tzinfo=timezone.utc)  # Garantir que o datetime seja aware
    cursor.execute(insert_query, (machine_id, sensor_id, timestamp, value))
    
    conn.commit()
    cursor.close()
    conn.close()
    return 0

def calculateMovingAverage(values: Deque[float]) -> float:
    if not values:
        return 0.0  # Se a fila estiver vazia, retorna 0 como média móvel

    total = sum(values)
    return total / len(values)

def calculateZScore(value: float, values: Deque[float]) -> float:
    if not values:
        return 0.0  # Se a fila estiver vazia, retorna 0 como Z-score

    mean = calculateMovingAverage(values)
    variance = sum((val - mean) ** 2 for val in values) / len(values)
    std_deviation = np.sqrt(variance)

    if std_deviation == 0.0:
        return 0.0  # Retorna 0 se o desvio padrão for zero para evitar divisão por zero

    return (value - mean) / std_deviation

def calculateTrend(values: Deque[float]) -> float:
    n = len(values)
    sum_x = 0.0
    sum_y = 0.0
    sum_xy = 0.0
    sum_xx = 0.0

    if n < 2:
        return 0.0

    for i in range(n):
        x = i + 1
        y = values[i]
        sum_x += x
        sum_y += y
        sum_xy += x * y
        sum_xx += x * x

    slope = (n * sum_xy - sum_x * sum_y) / (n * sum_xx - sum_x * sum_x)

    return slope

def process_sensor_data(machine_id, sensor_id, timestamp, value, sensorData, sensorTimestamp):
    print(f"\n\n--------------------->    Análise de dados para o sensor {sensor_id}   <---------------------")
    time = parser.parse(timestamp).replace(tzinfo=timezone.utc)  # Garantir que o datetime seja aware
    print(f"Data: {time.strftime('%d/%m/%Y, Hora: %H:%M:%S')}")
    print(f"ID da máquina: {machine_id}")

    sensorData.append(value)
    if len(sensorData) > MOVING_AVERAGE_WINDOW:
        sensorData.popleft()  # Usar popleft() para deque

    if sensorData:
        movingAverage = calculateMovingAverage(sensorData)
        print(f"Média móvel do uso de {sensor_id}: {movingAverage}")
        post_metric(machine_id, f"{sensor_id}.{sensor_id}_moving_average", timestamp, movingAverage)

        zScore = calculateZScore(value, sensorData)
        zScoreThreshold = 1.0

        if abs(zScore) > zScoreThreshold:
            print(f"{RED}[ALARME]{RESET} Outlier detectado: {value}")
            post_metric(machine_id, f"alarms.{sensor_id}_outlier", timestamp, 1)
        else:
            print(f"Uso normal de {sensor_id}: {value}")

        trend = calculateTrend(sensorData)
        print(f"Tendência do uso de {sensor_id}: {trend}")
        post_metric(machine_id, f"{sensor_id}.{sensor_id}_trend", timestamp, trend)
        print("----------------------------------------------------------------------------------------------")

cpuUsageData = deque()
def process_sensor_data_cpu(machine_id, sensor_id, timestamp, value):
    global cpuUsageData
    process_sensor_data(machine_id, sensor_id, timestamp, value, cpuUsageData, None)

memUsageData = deque()
def process_sensor_data_mem(machine_id, sensor_id, timestamp, value):
    global memUsageData
    process_sensor_data(machine_id, sensor_id, timestamp, value, memUsageData, None)

def check_sensor_inactivity(machine_id, sensor_id, sensor_name):
    expected_interval_seconds = 30
    max_expected_delay = expected_interval_seconds

    while True:
        time.sleep(20)
        current_time = datetime.utcnow().replace(tzinfo=timezone.utc)  # Garantir que o datetime seja aware
        with threading.Lock():
            last_timestamp = next((sensor['last_timestamp'] for sensor in monitored_sensors if sensor['sensor_id'] == sensor_id), None)
            if last_timestamp:
                last_time = parser.parse(last_timestamp).replace(tzinfo=timezone.utc)  # Garantir que o datetime seja aware
                seconds_since_last_timestamp = (current_time - last_time).total_seconds()

                if seconds_since_last_timestamp > max_expected_delay:
                    print(f"{RED}\n⚠️ - [ALARME] {RESET}Dados do sensor {sensor_name} da máquina {machine_id} não foram recebidos por mais de 10 períodos de tempo previstos.")
                    post_metric(machine_id, f"alarms.inactive_{sensor_name}", current_time.strftime("%Y-%m-%dT%H:%M:%S"), 1)

monitored_sensors = []

def add_monitored_sensor(sensor_id, timestamp):
    with threading.Lock():
        monitored_sensors.append({'sensor_id': sensor_id, 'last_timestamp': timestamp})

def is_sensor_monitored(sensor_id):
    with threading.Lock():
        return any(sensor['sensor_id'] == sensor_id for sensor in monitored_sensors)

def process_message(machine_id, sensor_id, timestamp, sensor_name):
    if not is_sensor_monitored(sensor_id):
        threading.Thread(target=check_sensor_inactivity, args=(machine_id, sensor_id, sensor_name)).start()
        add_monitored_sensor(sensor_id, timestamp)

class MqttCallback:
    def __init__(self, client):
        self.client = client

    def on_message(self, client, userdata, msg):
        j = json.loads(msg.payload.decode())
        topic = msg.topic.split('/')
        machine_id = topic[2]
        sensor_id = topic[3]
        timestamp = j["timestamp"]
        value = j["value"]

        post_metric(machine_id, f"{sensor_id}.{sensor_id}", timestamp, value)
        if sensor_id == "cpu_usage":
            process_sensor_data_cpu(machine_id, sensor_id, timestamp, value)
            process_message(machine_id, f"{sensor_id}{machine_id}", timestamp, sensor_id)
        else:
            process_sensor_data_mem(machine_id, sensor_id, timestamp, value)
            process_message(machine_id, f"{sensor_id}{machine_id}", timestamp, sensor_id)

def main():
    create_table_if_not_exists()

    client_id = "clientId"
    client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION1, client_id)

    callback = MqttCallback(client)
    client.on_message = callback.on_message

    client.connect(BROKER_ADDRESS)
    client.subscribe("/sensors/#", QOS)

    print("Subscribed")

    client.loop_start()

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("Exiting")
        client.loop_stop()
        client.disconnect()

if __name__ == "__main__":
    main()
